### PR TITLE
Fix ConcatType backward type inference

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##

Added backward type inference support for _rnn_param_concat.
fixes: https://github.com/apache/incubator-mxnet/issues/15716

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/concat.cc [hunk 1]
```diff
@@ -154,18 +155,29 @@ bool ConcatType(const nnvm::NodeAttrs& attrs,
...
+  // if out types are known in types are unknown
+  } else if ((*out_type)[0] != -1 && dtype == -1) {
+    in_type->clear();
+    for (size_t i = 0; i < nin; ++i) {
+      in_type->push_back((*out_type)[0]);
+    }
+  // if both out_types and in_types are known, and different
+  } else if ((*out_type)[0] != -1 && dtype != -1 && ((*out_type)[0] != dtype)) {
+    std::ostringstream os;
+    os << "Type inconsistent, Provided = "
```
Comment (ptrendx) [inline src/operator/nn/concat.cc]: "Provided output type" maybe?
Also, typically `LOG(FATAL)` was used for such errors - is this a new preferred way?

Comment (anirudh2290) [reply src/operator/nn/concat.cc]: within InferType logic InferTypeError is thrown for example: https://github.com/apache/incubator-mxnet/blob/master/src/operator/operator_common.h#L229

I was trying to keep it consistent with this. I can change the message to Provided output type

Comment (anirudh2290) [reply src/operator/nn/concat.cc]: I have changed the error message.

Comment (ptrendx) [reply src/operator/nn/concat.cc]: Makes sense, thanks for explanation :-)

**Pull request merged**